src/composite/reduce.jl

```diff
@@ -22,29 +22,56 @@ chsubblocks(x::Sum{N}, it) where N = Sum{N}(it)
 
 function apply!(r::AbstractRegister, x::Sum)
     isempty(x.list) && return r
+    length(x.list) == 1 && return apply!(r, x.list[])
 
-    out = copy(r)
-    apply!(out, first(x))
-    for k in 2:length(x)
-        out += apply!(copy(r), x[k])
-    end
-    copyto!(r, out)
-    return r
+    res = mapreduce(blk->apply!(copy(r), blk), +, x.list[1:end-1])
+    apply!(r, x.list[end])
+    r.state .+= res.state
+    r
 end
 
 export Sum
 
 subblocks(x::Sum) = x.list
 cache_key(x::Sum) = map(cache_key, x.list)
+Base.copy(x::Sum) = Sum(copy(x.list))
+Base.similar(c::Sum{N}) where N = Sum{N}(empty!(similar(c.list)))
+
+function Base.:(==)(lhs::Sum{N}, rhs::Sum{N}) where {N}
+    (length(lhs.list) == length(rhs.list)) && all(lhs.list .== rhs.list)
+end
+
+for FUNC in [:length, :iterate, :getindex, :eltype, :eachindex, :popfirst!, :lastindex]
+    @eval Base.$FUNC(x::Sum, args...) = $FUNC(subblocks(x), args...)
+end
+
+Base.getindex(c::Sum, index::Union{UnitRange, Vector}) = Sum(getindex(c.list, index))
```

Comment:
where is the element-wise `getindex`? why delete this and iteration method.

Reply:
See above
```julia
for FUNC in [:length, :iterate, :getindex, :eltype, :eachindex, :popfirst!, :lastindex]
    @eval Base.$FUNC(x::Sum, args...) = $FUNC(subblocks(x), args...)
end
```

Reply:
Already defined in
```julia
for FUNC in [:length, :iterate, :getindex, :eltype, :eachindex, :popfirst!, :lastindex]
    @eval Base.$FUNC(x::Sum, args...) = $FUNC(subblocks(x), args...)
end
```